pkg/controllers/deprovisioning/orchestration/queue.go

```diff
@@ -0,0 +1,325 @@
+		if !requeue {
+			// If the command timed out, log the last error.
+			if err != nil {
+				// If the command failed, bail on the action.
+				// 1. Emit metrics for launch failures
+				// 2. Ensure cluster state no longer thinks these nodes are deleting
+				// 3. Remove it from the Queue's internal data structure
+				deprovisioningReplacementNodeLaunchFailedCounter.WithLabelValues(cmd.Reason).Inc()
+				q.cluster.UnmarkForDeletion(lo.Map(cmd.Candidates, func(s *state.StateNode, _ int) string { return s.ProviderID() })...)
+				err = multierr.Append(err, q.setNodesUnschedulable(ctx, false, lo.Map(cmd.Candidates, func(s *state.StateNode, _ int) string { return s.Node.Name })...))
+				logging.FromContext(ctx).Debugf("handling deprovisioning command, %w, for nodes %s", err,
+					strings.Join(lo.Map(cmd.Candidates, func(s *state.StateNode, _ int) string {
+						return s.Name()
+					}), ","))
+			}
+			q.Remove(hash, cmd)
+			continue
+		}
+		q.RateLimitingInterface.Done(hash)
+		// Requeue command if not complete
+		q.RateLimitingInterface.AddRateLimited(hash)
+	}
+	logging.FromContext(ctx).Errorf("deprovisioning queue is broken and has shutdown")
+}
+
+// Add adds commands to the Queue
+// Each command added to the queue should already be validated and ready for execution.
+func (q *Queue) Add(cmd *Command) error {
+	providerIDs := lo.Map(cmd.Candidates, func(s *state.StateNode, _ int) string {
+		return s.ProviderID()
+	})
+	// First check if we can add the command.
+	if err := q.CanAdd(providerIDs...); err != nil {
+		return fmt.Errorf("adding command, %w", err)
+	}
+	hash, err := Hash(cmd)
+	if err != nil {
+		return err
+	}
+	for _, candidate := range cmd.Candidates {
+		q.candidateProviderIDToCommandID[candidate.ProviderID()] = hash
+		q.idToCommand[hash] = cmd
+	}
+	// Idempotently mark for deletion
+	q.cluster.MarkForDeletion(providerIDs...)
+	q.RateLimitingInterface.Add(hash)
+	return nil
+}
+
+// Remove fully clears the queue of all references of a hash/command
+func (q *Queue) Remove(hash uint64, cmd *Command) {
+	q.RateLimitingInterface.Forget(hash)
+	for _, candidate := range cmd.Candidates {
+		delete(q.candidateProviderIDToCommandID, candidate.ProviderID())
+	}
+	delete(q.idToCommand, hash)
+	q.RateLimitingInterface.Done(hash)
+}
+
+// CanAdd is a quick check to see if the candidate is already part of a deprovisioning action
+func (q *Queue) CanAdd(ids ...string) error {
+	var err error
+	for _, id := range ids {
+		if _, ok := q.candidateProviderIDToCommandID[id]; ok {
+			err = multierr.Append(err, fmt.Errorf("candidate is part of active deprovisioning decision"))
+		}
+	}
+	return err
+}
+
+// Handle will check for timeouts, then execute a wait or termination.
+// If there
+func (q *Queue) Handle(ctx context.Context, cmd *Command) (bool, error) {
+	// logging.FromContext(ctx).Infof("DEBUGGING: This is the current time: %s, this is the time added: %s", q.clock)
+	if q.clock.Since(cmd.TimeAdded) > maxRetryDuration {
```

Comment:
Does it make sense to let the caller handle this check? If you let the caller handle it, then you don't need to keep track of the `LastError` in the command. You can just return an error from here and then if the command is beyond the timeout, you just print the error and handle the failure path

Reply:
I think we still need to maintain a separate `LastError`. For a given iteration, we should only return a timeout error when it's actually been past the 10 minutes. Once it's timed out, we probably shouldn't check again or make any other commands, but we always could. Using the `LastError` means that we're for sure only considering the errors for a 10 minute period, but if we allow the last iteration to occur past the 10 minutes, then we could propagate the most recent error like that.  What are your thoughts?